### PR TITLE
gnome-base/gnome-session: Version bump to 46.0

### PR DESCRIPTION
**Pull request by flint2**
elogind support Work In Progress

### CODE REVIEW

**Comment by erikmack** [comment]
Can the unused eclass be removed?

**Comment by pacho2** [comment]
Hello,

It seems that the way to go could be the same as followed by LFS, Void and Devuan:
https://www.linuxfromscratch.org/blfs/view/svn/gnome/gnome-session.html
(Command explanations section)
https://github.com/void-linux/void-packages/blob/master/srcpkgs/gnome-session/template

In summary: 
When elogind USE is enabled, it should pass -Dsystemd_journal=false and some dummy value for -Dsystemduserunitdir. In LFS they pass /tmp, in Void they pass the regular /usr/lib/systemd/user (is that even used in elogind systems?).

Once that is passed, build should work, and elogind should be used at runtime (RDEPEND).

Sadly, I don't have any elogind system... then, we should try and see if it works for those users I think :/

**Comment by robszy** [comment]
I have elogind on my system but i don't think this will resolve the issue. 

Real problem is hardcoded systemd headers in source and not added dependency for elogind.

**Comment by flint2** [comment]
Removing elogind and making systemd a hard dependency as per upstream https://gitlab.gnome.org/GNOME/gnome-session/-/commit/76534bcc5e4e0ee38b8541dbb413d4b36d30d9d7

**Comment by robszy** [comment]
We need patch for elogind support to resolve this issue but from what i see is not very complicated :)

**Comment by pacho2** [comment]
> We need source patching for elogind support to resolve this issue but from what i see is not very complicated :)

Are you able to test on an elogind system?

What happens when you try to install elogind and pass:
-Dsystemd_journal=false and -Dsystemduserunitdir=/tmp (or any other void location)
?

**Comment by flint2** [comment]
> > We need source patching for elogind support to resolve this issue but from what i see is not very complicated :)
> 
> Are you able to test on an elogind system?
> 
> What happens when you try to install elogind and pass: -Dsystemd_journal=false and -Dsystemduserunitdir=/tmp (or any other void location) ?

This is in a chrooted environment with non systemd profile (  default/linux/amd64/23.0 (stable)  ) :
- installed elogind
added -Dsystemd_journal=false   --> error in configure stage:

``` 
>>> Configuring source in /var/tmp/portage/gnome-base/gnome-session-46.0/work/gnome-session-46.0 ...
meson setup -Db_lto=false --libdir lib64 --localstatedir /var/lib --prefix /usr --sysconfdir /etc --wrap-mode nodownload --build.pkg-config-path /usr/share/pkgconfig --pkg-config-path /usr/share/pkgconfig --native-file /var/tmp/portage/gnome-base/gnome-session-46.0/temp/meson.x86_64-pc-linux-gnu.amd64.ini -Db_pch=false -Dwerror=false -Dbuildtype=plain -Ddeprecation_flags=false -Dsession_selector=true -Ddocbook=false -Dman=true -Dsystemd_journal=false -Dsystemduserunitdir=/tmp /var/tmp/portage/gnome-base/gnome-session-46.0/work/gnome-session-46.0 /var/tmp/portage/gnome-base/gnome-session-46.0/work/gnome-session-46.0-build
The Meson build system
Version: 1.4.1
Source dir: /var/tmp/portage/gnome-base/gnome-session-46.0/work/gnome-session-46.0
Build dir: /var/tmp/portage/gnome-base/gnome-session-46.0/work/gnome-session-46.0-build
Build type: native build

meson.build:1:0: ERROR: Unknown options: "systemd_journal"

A full log can be found at /var/tmp/portage/gnome-base/gnome-session-46.0/work/gnome-session-46.0-build/meson-logs/meson-log.txt
```
This option was removed in https://gitlab.gnome.org/GNOME/gnome-session/-/commit/76534bcc5e4e0ee38b8541dbb413d4b36d30d9d7





Tested with removed -Dsystemd_journal=false --> error in configure phase


``` 
>>> Configuring source in /var/tmp/portage/gnome-base/gnome-session-46.0/work/gnome-session-46.0 ...
meson setup -Db_lto=false --libdir lib64 --localstatedir /var/lib --prefix /usr --sysconfdir /etc --wrap-mode nodownload --build.pkg-config-path /usr/share/pkgconfig --pkg-config-path /usr/share/pkgconfig --native-file /var/tmp/portage/gnome-base/gnome-session-46.0/temp/meson.x86_64-pc-linux-gnu.amd64.ini -Db_pch=false -Dwerror=false -Dbuildtype=plain -Ddeprecation_flags=false -Dsession_selector=true -Ddocbook=false -Dman=true -Dsystemduserunitdir=/tmp /var/tmp/portage/gnome-base/gnome-session-46.0/work/gnome-session-46.0 /var/tmp/portage/gnome-base/gnome-session-46.0/work/gnome-session-46.0-build
The Meson build system
Version: 1.4.1
Source dir: /var/tmp/portage/gnome-base/gnome-session-46.0/work/gnome-session-46.0
Build dir: /var/tmp/portage/gnome-base/gnome-session-46.0/work/gnome-session-46.0-build
Build type: native build
Project name: gnome-session
Project version: 46.0
C compiler for the host machine: x86_64-pc-linux-gnu-gcc (gcc 13.3.1 "x86_64-pc-linux-gnu-gcc (Gentoo 13.3.1_p20240614 p17) 13.3.1 20240614")
C linker for the host machine: x86_64-pc-linux-gnu-gcc ld.bfd 2.42
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: YES (/usr/bin/x86_64-pc-linux-gnu-pkg-config) 2.2.0
Run-time dependency gio-2.0 found: YES 2.78.6
Run-time dependency glib-2.0 found: YES 2.78.6
Found CMake: /usr/bin/cmake (3.28.5)
Run-time dependency gtk+-3.0 found: NO (tried cmake)

meson.build:84:10: ERROR: Dependency lookup for gtk+-3.0 with method 'pkgconfig' failed: Could not generate cflags for gtk+-3.0:
Package libsystemd was not found in the pkg-config search path.
Perhaps you should add the directory containing `libsystemd.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libsystemd', required by 'dbus-1', not found

```
This dependency was added in https://gitlab.gnome.org/GNOME/gnome-session/-/commit/9dd66a51a2e0623d716b8b9907af374b40f6b6b6

**Comment by pacho2** [comment]
> This dependency was added in https://gitlab.gnome.org/GNOME/gnome-session/-/commit/9dd66a51a2e0623d716b8b9907af374b40f6b6b6

```
# Check for systemd
systemd_userunitdir = get_option('systemduserunitdir')
if systemd_userunitdir == ''
  systemd_dep = dependency('systemd', version: '>= 242', required: true)
  systemd_userunitdir = systemd_dep.get_variable(pkgconfig: 'systemduserunitdir',
                                                 pkgconfig_define: ['prefix', session_prefix])
endif

libsystemd_dep = dependency('libsystemd', version: '>= 209', required: true)
session_bin_deps += libsystemd_dep

```

It seems elogind on Void and LFS is able to provide "libsystemd" because they create some links:
https://www.linuxfromscratch.org/blfs/view/svn/general/elogind.html
https://github.com/void-linux/void-packages/blob/master/srcpkgs/elogind/template#L52

But I don't know if we could do the same, maybe @a17r will know if that would be possible

**Comment by flint2** [comment]
> Maybe this could work too:
> 
> ```
> --- meson.build.old	2024-07-29 16:43:01.812864903 +0200
> +++ meson.build	2024-07-29 16:54:16.654741387 +0200
> @@ -113,8 +113,13 @@
>                                                   pkgconfig_define: ['prefix', session_prefix])
>  endif
>  
> -libsystemd_dep = dependency('libsystemd', version: '>= 209', required: true)
> -session_bin_deps += libsystemd_dep
> +libsystemd_dep = dependency('libsystemd', version: '>= 209', required: false)
> +if libsystemd_dep.found()
> +  session_bin_deps += libsystemd_dep
> +else
> +  elogind_dep dependency('libelogind', version: '>=209', required: true)
> +  session_bin_deps += elogind_dep
> +fi
>  
>  configure_file(
>    output: 'config.h',
> ```

Added patch with minor fixes.

Even with this change
```
-               -Dsystemduserunitdir="$(systemd_get_userunitdir)"
+               -Dsystemduserunitdir="/tmp"
```

Now is failing in configure phase with:

```
>>> Configuring source in /var/tmp/portage/gnome-base/gnome-session-46.0/work/gnome-session-46.0 ...
meson setup -Db_lto=false --libdir lib64 --localstatedir /var/lib --prefix /usr --sysconfdir /etc --wrap-mode nodownload --build.pkg-config-path /usr/share/pkgconfig --pkg-config-path /usr/share/pkgconfig --native-file /var/tmp/portage/gnome-base/gnome-session-46.0/temp/meson.x86_64-pc-linux-gnu.amd64.ini -Db_pch=false -Dwerror=false -Dbuildtype=plain -Ddeprecation_flags=false -Dsession_selector=true -Ddocbook=false -Dsystemduserunitdir=/usr/lib/systemd/user -Dman=true /var/tmp/portage/gnome-base/gnome-session-46.0/work/gnome-session-46.0 /var/tmp/portage/gnome-base/gnome-session-46.0/work/gnome-session-46.0-build
The Meson build system
Version: 1.5.1
Source dir: /var/tmp/portage/gnome-base/gnome-session-46.0/work/gnome-session-46.0
Build dir: /var/tmp/portage/gnome-base/gnome-session-46.0/work/gnome-session-46.0-build
Build type: native build
Project name: gnome-session
Project version: 46.0
C compiler for the host machine: x86_64-pc-linux-gnu-gcc (gcc 13.3.1 "x86_64-pc-linux-gnu-gcc (Gentoo 13.3.1_p20240614 p17) 13.3.1 20240614")
C linker for the host machine: x86_64-pc-linux-gnu-gcc ld.bfd 2.42
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: YES (/usr/bin/x86_64-pc-linux-gnu-pkg-config) 2.2.0
Run-time dependency gio-2.0 found: YES 2.78.6
Run-time dependency glib-2.0 found: YES 2.78.6
Found CMake: /usr/bin/cmake (3.30.1)
Run-time dependency gtk+-3.0 found: NO (tried cmake)

meson.build:84:10: ERROR: Dependency lookup for gtk+-3.0 with method 'pkgconfig' failed: Could not generate cflags for gtk+-3.0:
Package libsystemd was not found in the pkg-config search path.
Perhaps you should add the directory containing `libsystemd.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libsystemd', required by 'dbus-1', not found


```

**Comment by pacho2** [comment]
> Package libsystemd was not found in the pkg-config search path.
> Perhaps you should add the directory containing `libsystemd.pc'
> to the PKG_CONFIG_PATH environment variable
> Package 'libsystemd', required by 'dbus-1', not found
> ```

Is your sys-apps/dbus compiled with elogind support? Otherwise /usr/lib64/pkgconfig/dbus-1.pc will have:
```
Requires.private: libsystemd >= 209
```

**Comment by pacho2** [comment]
You will also need gnome-base/gnome-desktop:3 with systemd USE flag disabled

**Comment by flint2** [comment]
Now is compiling with elogind

File: gnome-base/gnome-session/gnome-session-46.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+src_configure() {
+	local emesonargs=(
+		-Ddeprecation_flags=false
+		-Dsession_selector=true # gnome-custom-session
+		$(meson_use doc docbook)
+		-Dman=true
+	)
+	use systemd && local emesonargs+=( -Dsystemduserunitdir="$(systemd_get_userunitdir)" )
+	use elogind && local emesonargs+=( -Dsystemduserunitdir="/tmp" )
```
Comment (pacho2) [inline gnome-base/gnome-session/gnome-session-46.0.ebuild:81]: Supposedly, we can set here any value... maybe we could set something under ${WORKDIR}. That file wouldn't finally be installed and, like that, we don't pollute /tmp.

Other option could be to simply install it in the usual $(systemd_get_userunitdir); gvfs ebuild does that and it seems it won't bother elogind.

Comment (pacho2) [reply gnome-base/gnome-session/gnome-session-46.0.ebuild:81]: Maybe you can simply rely on -Dsystemduserunitdir="${T}"

**Comment by pacho2** [comment]
What did happen with:
newmenu "${FILESDIR}/defaults.list-r6" gnome-mimeapps.list
?

It seems it was dropped

**Comment by flint2** [comment]
> What did happen with: newmenu "${FILESDIR}/defaults.list-r6" gnome-mimeapps.list ?
> 
> It seems it was dropped

Added back in https://github.com/gentoo/gentoo/pull/34748/commits/6a5d6b526d8948a72c85c566462bccb4f9773c83

**Comment by robszy** [comment]
@flint2 How it handles includes like #include <systemd/sd-journal.h> because it should have elogind/sd-journal.h in  /usr/include/elogind/sd-journal.h. You have systemd also installed in chroot so you have include files ?

**Comment by flint2** [comment]
> @flint2 How it handles includes like #include <systemd/sd-journal.h> because it should have elogind/sd-journal.h in /usr/include/elogind/sd-journal.h. You have systemd also installed in chroot so you have include files ?

This is the way

```
gjoandet@gentoorc ~ $ locate sd-journal.h
/usr/include/elogind/sd-journal.h
/usr/include/elogind/systemd/sd-journal.h
gjoandet@gentoorc ~ $ equery b /usr/include/elogind/sd-journal.h
 * Searching for /usr/include/elogind/sd-journal.h ... 
sys-auth/elogind-252.9-r2 (/usr/include/elogind/sd-journal.h -> systemd/sd-journal.h)
sys-auth/elogind-252.9-r2 (/usr/include/elogind/systemd/sd-journal.h)
gjoandet@gentoorc ~ $ equery u gnome-session
[ Legend : U - final flag setting for installation]
[        : I - package is installed with flag     ]
[ Colors : set, unset                             ]
 * Found these USE flags for gnome-base/gnome-session-46.0:
 U I
 - - doc     : Add extra documentation (API, Javadoc, etc). It is recommended to enable per package instead of globally
 + + elogind : Support sys-auth/elogind for session tracking
 - - systemd : Enable use of systemd-specific libraries and features like socket activation or session tracking
gjoandet@gentoorc ~ $ eselect profile show
Current /etc/portage/make.profile symlink:
  default/linux/amd64/23.0/desktop

```
And no, systemd is not installed, and now I use a full amd64 VM with multilib openrc to test and gnome is working OK

**Comment by flint2** [comment]
Now with latest ebuilds is failing:

```
FAILED: plugins/sharing/gsd-sharing.p/gsd-sharing-manager.c.o 
x86_64-pc-linux-gnu-gcc -Iplugins/sharing/gsd-sharing.p -Iplugins/sharing -I../gnome-settings-daemon-46.0/plugins/sharing -I. -I../gnome-settings-daemon-46.0 -Iplugins/common -I../gnome-settings-daemon-46.0/plugins/common -Ignome-settings-daemon -I../gnome-settings-daemon-46.0/gnome-settings-daemon -I/usr/include/gio-unix-2.0 -I/usr/lib64/libffi/include -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libpng16 -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -DHAVE_CONFIG_H -march=native -O2 -pipe -pthread '-DG_LOG_DOMAIN="sharing-plugin"' '-DPLUGIN_NAME="sharing"' '-DPLUGIN_DBUS_NAME="org.gnome.SettingsDaemon.Sharing"' '-DGNOME_SETTINGS_LOCALEDIR="/usr/share/locale"' -MD -MQ plugins/sharing/gsd-sharing.p/gsd-sharing-manager.c.o -MF plugins/sharing/gsd-sharing.p/gsd-sharing-manager.c.o.d -o plugins/sharing/gsd-sharing.p/gsd-sharing-manager.c.o -c ../gnome-settings-daemon-46.0/plugins/sharing/gsd-sharing-manager.c
../gnome-settings-daemon-46.0/plugins/sharing/gsd-sharing-manager.c:23:10: fatal error: systemd/sd-login.h: No such file or directory
   23 | #include <systemd/sd-login.h>
      |          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

Need to check again

**Comment by flint2** [comment]
Problem is with gsd-sharing

After disabling plugin it compiles
```
--- /var/tmp/portage/gnome-base/gnome-settings-daemon-46.0/work/gnome-settings-daemon-46.0/plugins/meson.build  2024-03-16 19:14:58.000000000 -0300
+++ /tmp/meson.build    2024-08-04 18:56:11.278773756 -0300
@@ -7,7 +7,6 @@
     ['keyboard', 'Keyboard', 'GNOME keyboard configuration'],
     ['media-keys', 'MediaKeys', 'GNOME keyboard shortcuts'],
     ['screensaver-proxy', 'ScreensaverProxy', 'GNOME FreeDesktop screensaver'],
-    ['sharing', 'Sharing', 'GNOME file sharing'],
     ['sound', 'Sound', 'GNOME sound sample caching'],
     ['usb-protection', 'UsbProtection', 'GNOME USB protection'],
     ['xsettings', 'XSettings', 'GNOME XSettings'],
```

Needs more analysis

**Comment by flint2** [comment]
Please follow up on https://bugs.gentoo.org/937244

**Comment by robszy** [comment]
@flint2  It is very wierd. It should work as you've said with include #include <systemd/sd-login.h> ?

**Comment by flint2** [comment]
> @flint2  It is very wierd. It should work as you've said with include #include <systemd/sd-login.h> ?
> 
> EDIT: very simple; doesn't have include to /usr/include/elogind :)
> And this is another package gsd so it should be in other pr

Check this https://github.com/gentoo/gentoo/pull/37960

**Pull request closed without merging**